`General`: Add type filtering to average course scores chart for instructors

willergeorg opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [ ] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
This PR changes some existing functionality and adds some details to the average score chart on course statistics page
### Description
<!-- Describe your changes in detail -->

- **Data label**: Above the chart bars, a data label now displays now the represented average exercise score numerically.
- **Content order**: The exercises are now ordered ascending in their average score.
- **Dynamic chart coloring**: Previously, all bars representing an average score >= 90% were colored green and all bars representing an average score <= 25% were colored red. We decided to change this logic. Now, the in average lowest performing third of the course exercises is colored red and the in average best performing third is colored green. Note, that the exercises might not be necessarily separated in exact thirds. This depends on the general amount of exercises and the average score distribution. If for example 5 out of 10 exercises share the lowest score, it would not make much sense to declare 3 of them as lowest third and 2 of them as part of the average third. 
- **Color legend**: A color legend is added in order to explain the color code. By clicking on an arbitrary legend entry, filtering of the chart is possible:
     - When selecting an interval that is already selected, all other selected intervals get deselected
     - When selecting an interval that is currently the only one selected, all intervals get selected again
     - When selecting an interval that is currently deselected, it is added to the current filter
- **Type filter**: Additionally, it is now possible to filter the chart by exercise types.

Additional change:

- I changed the page structure a bit and moved the tab switcher manipulating the displayed time span below the average scores chart and separated it with a light grey border from the other charts. I think this reduces the confusion of the user since the current layout might imply that the average scores chart is time span sensitive as well and can be manipulated by the tab switcher (which is not the case apparently).

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Instructor
- 1 Course with some contributions

1. Log in to Artemis
2. Navigate to Course Administration
3. Select "Statistics" of the corresponding course and inspect the first chart
4. Make sure that the exercises are ordered by their average score
5. Play around with the filter options (Type and interval filter). Make sure both work as expected. Note, that the interval filter will always reset the type filter. This is done in order to prevent undesired side effects. 

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Please use the schema "Branches % | Lines %" -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
<!-- - ExerciseService.java: 85% | 77% -->
<!-- - programming-exercise.component.ts: 13% | 95% -->
| File name | Branch coverage (%) | Line coverage (%) |
| -- | -- | -- |
| `exercise-scores-chart.component.ts` | 79.41 | 98 |
| `chart-exercise-type-filter.directive.ts` | 50 | 100 |
| `statistics-average-scores-graph.component.ts` | 79.41 | 95.86 |
### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
![image](https://user-images.githubusercontent.com/80622272/154841722-64b436a0-f035-49db-b057-c1abe621f271.png)

coaxial1 commented on src/main/webapp/app/course/manage/course-management-statistics.component.html:
```diff
@@ -1,6 +1,18 @@
+        [courseId]="courseId"
+    >
+    </jhi-statistics-average-score-graph>
+    <div style="border-top: 3px solid lightgray" class="w-100 mb-4"></div>
```
Should we rather avoid inline styles here? Especially inline colors are bad when it comes to theming 😬

willergeorg replied:
You are right in general. But since there is no dedicated scss file for this component, creating one only for this separator seemed a bit disproportional for me. But if this would be convenient for theming, I will add it of course.

coaxial1 commented on src/main/webapp/app/shared/chart/chart-exercise-type-filter.directive.ts:
```diff
@@ -0,0 +1,47 @@
+    protected initializeFilterOptions(exerciseScores: any[]): void {
+        this.typeSet = new Set(exerciseScores.map((score) => score.exerciseType));
+        this.typeSet.forEach((type) => {
+            this.chartFilter.set(type.toLowerCase().replace('_', '-'), true);
```
The chain ``toLowerCase().replace('_', '-')`` is used three times in this file. I suggest to extract it into a static function instead, or just encapsulate all access to ``this.chartFilter`` into ``getChartFilter`` and ``setChartFilter`` methods (or named similarly) that include these string operations

coaxial1 commented on src/main/webapp/app/shared/statistics-graph/statistics-average-score-graph.component.html:
```diff
@@ -36,11 +40,46 @@ <h4>{{ 'artemisApp.courseStatistics.courseAverage' | artemisTranslate }}: {{ cou
+                </li>
+                <li *ngFor="let type of typeSet">
+                    <label>
+                        <input class="ms-2" (change)="toggleType(type)" [checked]="chartFilter.get(type.toLowerCase().replace('_', '-'))" type="checkbox" />
```
This refers to my previous comment: You could reuse a new ``getChartFilter`` method here as well

coaxial1 reviewed (changes requested):
Code looks good, just added a few remarks.

Additionally, after testing this on TS1, I'd like to suggest to provide an easier way to get back to selecing all "performance thirds" after having selected one. This must not necessarily be in this PR, though..
For example, if I select the "Lowest performing third", everything works, but there is no easy way to revert this selection and see all thirds again other than selecting the other two one by one.
A solution could be that if someone selects a third and then clicks it again (while it's the only selected one), it goes back to selecting all instead.

willergeorg commented:
> Code looks good, just added a few remarks.
> 
> Additionally, after testing this on TS1, I'd like to suggest to provide an easier way to get back to selecing all "performance thirds" after having selected one. This must not necessarily be in this PR, though.. For example, if I select the "Lowest performing third", everything works, but there is no easy way to revert this selection and see all thirds again other than selecting the other two one by one. A solution could be that if someone selects a third and then clicks it again (while it's the only selected one), it goes back to selecting all instead.

Thanks for your feedback. I incorporated it in 34629c9.

coaxial1 commented on src/main/webapp/app/overview/visualizations/exercise-scores-chart/exercise-scores-chart.component.html:
```diff
@@ -11,7 +11,7 @@ <h3>{{ 'artemisApp.exercise-scores-chart.title' | artemisTranslate }}</h3>
             <li *ngFor="let type of typeSet">
                 <label>
-                    <input class="ms-2" (change)="toggleExerciseType(type)" [checked]="chartFilter.get(type.toLowerCase().replace('_', '-'))" type="checkbox" />
+                    <input class="ms-2" (change)="toggleType(type)" [checked]="chartFilter.get(type.toLowerCase().replace('_', '-'))" type="checkbox" />
```
Found one more occurrence that could be replaced (right?) 😬

willergeorg replied:
Apparently I did not remove the file filter when searching for the command in all files. Thanks for the heads up. Changed in 20b96ce 👍

ole-ve reviewed (commented):
I think that I found an error while testing on TS1 with User 1 for [Johannes Stöhr Test Course](https://artemistest.ase.in.tum.de/course-management/511/course-statistics):
![Bildschirmfoto 2022-02-21 um 11 54 29](https://user-images.githubusercontent.com/73833780/154941426-2257cfd5-db23-4016-b891-9a462764a0b7.png)
I have selected the lowest-third interval, but an exercise from the upper-third interval is being displayed. The exercise is also displayed when filtering for the upper-third part.

ole-ve commented on src/main/webapp/app/shared/statistics-graph/statistics-average-score-graph.component.ts:
```diff
@@ -153,4 +178,173 @@ export class StatisticsAverageScoreGraphComponent implements OnInit {
+        let filterFunction;
+        switch (interval) {
+            case PerformanceInterval.LOWEST: {
+                filterFunction = (model: CourseManagementStatisticsModel) => model.averageScore <= this.weakestThirdUpperBoundary;
```
I assume that this may be related to the fact, that this course only contains one exercise. When there is only one exercise, the average score of all exercises equals the average score for this particular exercise. As you also allow equality ('=') in the cases for both LOWEST and BEST when comparing the averages, in both cases the exercise will be shown when filtering for lowest or best third interval.

willergeorg replied:
I changed the implementation in that way, that the thresholds are now exclusive. I think this should fix the problem.
![image](https://user-images.githubusercontent.com/80622272/155127282-1a84a592-6eb2-43a1-9e1f-7888bd2c6c78.png)

ole-ve replied:
Works as described, thanks.

ole-ve reviewed (commented):
Tested on TS1, I found another issue:

https://user-images.githubusercontent.com/73833780/155131107-bb650432-49ea-4029-b5ab-2ed22dfdbff3.mov


When selecting two intervals, disabling all filters and then also select the third interval, only exercises within this third interval are displayed.

ole-ve reviewed (commented):
I think this is how this error occurs:
1. Select the two intervals -> `exercisesScoresFilteredByPerformanceInterval` contains the exercises within both intervals (togglePerformanceInterval 2x times)
2. De-selecting the filter for an exercise type -> `currentlyDisplayableExercises` is set to empty (toggleType)
3. Selecting the third interval: the exercises in the third interval are added to `exercisesScoresFilteredByPerformanceInterval` and to `currentlyDisplayableExercises`.
Here, the problem occurs. In line 276, `exercisesScoresFilteredByPerformanceInterval` is assigned with the `currentlyDisplayableExercises` which only contains the exercises within the third interval (togglePerformanceInterval). Therefore, only exercises in this interval are shown.

willergeorg commented:
> I think this is how this error occurs:
> 
> 1. Select the two intervals -> `exercisesScoresFilteredByPerformanceInterval` contains the exercises within both intervals (togglePerformanceInterval 2x times)
> 2. De-selecting the filter for an exercise type -> `currentlyDisplayableExercises` is set to empty (toggleType)
> 3. Selecting the third interval: the exercises in the third interval are added to `exercisesScoresFilteredByPerformanceInterval` and to `currentlyDisplayableExercises`.
>    Here, the problem occurs. In line 276, `exercisesScoresFilteredByPerformanceInterval` is assigned with the `currentlyDisplayableExercises` which only contains the exercises within the third interval (togglePerformanceInterval). Therefore, only exercises in this interval are shown.

Good catch, I'm starting to realize that handling two separate filters is quite messy. My changes in c0c8799 should resolve the problem.

ole-ve reviewed (dismissed):
Thanks for implementing my feedback. Tested on TS1, works now as described. 👍

dveljacic reviewed (approved):
Tested on TS2. No unexpected behavior was found. ✅

willergeorg commented:
I only resolved a merge conflict. No implementation logic has been changed, therefore the reviews are still valid.

cremertim reviewed (approved):
tested on ts1. Code changes lgtm

valeriebucher reviewed (approved):
The order is correct and filter options are working as well. Looks good👍

canberkanar commented:
Tested on ts2. Everything works as expected and I was not able to replicate the issue reported by @ole-ve. Nicely done. 👍🏻

ole-ve reviewed (approved):
Re-approve code & manual testing
